### DISCUSSION TITLE
Use custom Kafka docker image

### DISCUSSION THREAD

**Discussion by ThePoipoi** [strimzi/strimzi-kafka-operator#5980] [Q&A] [answered]
Hi

Long time no see.
I'm opening this thread for a simple question: how can we use custom Kafka docker image ?
I've read the documentation, saying that we can specify the image to use, but it's not working as expected: both zookeeper and kafka pods are running, but the entity-operator one doesn't work.
We have this requirement, cause we want to use our own Kafka image, but deploy it with Strimzi (to avoid more deployment dev'). To test this scenario, I have used your _make build_ to generate docker images (everything went fine), and try to use one of the Kafka (2.8.0) image, renamed on my own.

My Kafka configuration:
![image](https://user-images.githubusercontent.com/25198074/144455751-27e82bfd-0f99-44dc-af3e-cc35f4f8a601.png)
![image](https://user-images.githubusercontent.com/25198074/144456420-47941c29-e3c0-4bef-af3b-e491e60b9905.png)

Describe on the cluster-entity-operator that failing (_CrashLoopBackOff_):
![image](https://user-images.githubusercontent.com/25198074/144456166-93c7d160-6ed1-4c24-a2d1-54871e64f400.png)

Moreover, I have a weird behavior: sometime, the quay.io/strimzi/kafka image is downloaded by something, but not always. I remove it between each deployment, and it come back sometimes.
![image](https://user-images.githubusercontent.com/25198074/144455455-ede3a21b-86f0-436e-9ca9-9ed5153fe9c4.png)

If I don't specify the image for zookeeper, topicOperator or userOperator, the Strimzi image is always downloaded during deployment. But when I do, it's random...

So, what I'm doing wrong ? Do you have any idea on why the cluster-entity-operator keep failing even if it's (in fact) your own Kafka docker image, just renamed with a custom name and specified by the "image" attribute ?
Thanks a lot :)

**Comment by scholzj** [comment, marked as answer]
The User and Topic operators are not using the Kafka images but the Operator images. So you are IMHO configuring it right, just to a bad value => you should use something like `antoine/operator:0.25.0` for example. 

Since you use node port listener, there will be also an init container being used => that might be one of the things pulling the Strimzi image. You should be able to configure it using the `brokerRackInitImage` option: https://strimzi.io/docs/operators/latest/full/using.html#property-kafka-brokerRackInitImage-reference ... again, this should be the operator and not the Kafka image.

BTW: Maybe you know it - but just in case you don't. You can configure the images like this in the `Kafka` custom resource. But you can also configure them centrally in the Cluster Operator deployment  and its environment variables: https://github.com/strimzi/strimzi-kafka-operator/blob/main/install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml#L50-L91 ... this is normally easier unless you want every Kafka cluster to use different images.

**Reply by ThePoipoi** [reply]
Haaaaaan, ok, I didn't know that.
Thanks a lot for those information, I'll try to work on both links.

To be more clear about our requirement: we try to deliver a solution that can be both deploy in an environment with or without a Strimzi Operator already installed, but using our own (customized) kafka image.
If the customer have nothing, our solution will deploy both Strimzi Operator and our custom Kafka image.
Itf it's already here, and the customer already have others Kafka clusters running, we want to use the existing Operator and just deploy a new Kafka cluster, but still with our custom image.
In both case, our customers want to be sure they are using our Kafka image and not your.

Do you think it can be done ?

**Reply by scholzj** [reply]
Well, as I said -> the Kafka CR defines the images only for this particular Kafka cluster. So you can use it to control the images just for it. So it should work. One think to keep in mind is that the images are coupled with the operator. So with operator 0.25, you should use your 0.25 images. But if you have operator 0.26, you would need 0.26 images.

**Reply by ThePoipoi** [reply]
Ok perfect, I'll try it and let you know.
Thanks again 👍

**Reply by ThePoipoi** [reply]
Hi Jakub.

Two different things today :

1) I don't succeed to use `brokerRackInitImage`. I don't know why, but every time I set it (to my custom image), Kafka pods refuse to start

![image](https://user-images.githubusercontent.com/25198074/145243324-a1952721-9502-4a1b-9be1-cd62b2ae3407.png)

![image](https://user-images.githubusercontent.com/25198074/145243108-e50f20c3-08f6-4f62-baf9-75f197232ead.png)

Does it mean that I've failed the Kafka Image build ? It's weird, cause without this parameter, everything works fine.

2) finally, it's the entity-operator itself that always download and use the quay.io/strimzi/kafka image

![image](https://user-images.githubusercontent.com/25198074/145242759-4f0ac6a5-250a-42d9-a79c-9f2034d02122.png)

So whatever I do, I'll always have 2 kafka images.
Am I doing something wrong ?

**Reply by scholzj** [reply]
Re 1) As I said before - the `brokerRackInitImage` is the operator image and not the Kafka image. So you are basically setting it to a wrong image and that is why it doesn't start.

Re 2) That is probably the TLS sidecar image? https://strimzi.io/docs/operators/latest/full/using.html#type-TlsSidecar-reference

**Reply by ThePoipoi** [reply]
I'm really sorry, I really need to clean my glasses...so, ok for point 1).
But as explained, we want to deliver a Kafka docker image without delivering the Operator's one, cause it can be already available and used on the customer environment. So we will never promote an image like `"antoine/operator:X.X.X"`
I guess there is no need to use `brokerRackInitImage` in this case, cause for me it's only the entity-operator that download and use the Strimzi image (I've checked it multiple times: the image is only downloaded when the entity-operator start, not before, nor by kafka or zookeeper images).

And obviously, thanks for your answer :)

**Reply by scholzj** [reply]
> But as explained, we want to deliver a Kafka docker image without delivering the Operator's one, cause it can be already available and used on the customer environment. So we will never promote an image like "antoine/operator:X.X.X"

Ah, ok. Well, in that case you should not set the `brokerRackInitImage` and let it use the original operator image from Strimzi. There is no way to use the Kafka image for it. But the TLS sidecars use the Kafka image -> the one in Entity Operator I pointed out and if you use Cruise Control, there is another TLS sidecar.
